### ISSUE TITLE
Google via OAuth

### ISSUE THREAD

**Issue by paulgessinger** [joeyates/imap-backup#79]
As far as I understand it, Google will phase out app passwords for G Suite (basically Gmail with a custom domain) for “security reasons”: https://gsuiteupdates.googleblog.com/2019/12/less-secure-apps-oauth-google-username-password-incorrect.html?m=1

Any idea how large the effort would be to support Google’s OAuth for accessing these types of accounts?

**Comment by joeyates** [comment]
@paulgessinger Thanks for letting me know.

Doing OAuth would require opening a browser and asking the user to go through the authorization flow.

it would certainly require a bit of work, but the main problem is that it means using this library with a cron job would become more difficult.

Here's an blog post by Martin Fowler about a similar use case: https://martinfowler.com/articles/command-line-google.html

Momentarily, seeing as there's a while before this gets enforced, I'll see if someone comes up with a PR

**Comment by paulgessinger** [comment]
Right, makes sense. In principle, some of this flow could be pretty manual I guess. There doesn’t need to be a fully automated auth. flow, as also described in the link you included. 

Once that’s done (i.e. once the refresh token is in the config file), everything else should not be problematic in a cron job, right? You just need to make sure the auth sequence ran once before. 

I have no experience in ruby, otherwise I’d have stab at it myself.

**Comment by paulgessinger** [comment]
I dug around a little bit. I found [this](https://github.com/nfo/gmail_xoauth) which mentions using Google's own [oauth2.py](https://github.com/google/gmail-oauth2-tools) tool to do the sequence with the google api. Subsequently, it should be possible to do something like

```ruby
require 'gmail_xoauth'
imap = Net::IMAP.new('imap.gmail.com', 993, usessl = true, certs = nil, verify = false)
imap.authenticate('XOAUTH2', 'myemail@gmail.com', my_oauth2_token)
messages_count = imap.status('INBOX', ['MESSAGES'])['MESSAGES']
puts "Seeing #{messages_count} messages in INBOX"
```
(copied verbatim from the first url) to get an imap connection, which I believe this tool operates off of.

**Comment by dankrad** [comment]
This app is now blocked by Google as a "Less secure app" :( so this enhancement might be necessary

**Comment by joeyates** [comment]
@paulgessinger If you're willing to beta test, I've implemented this in a branch https://github.com/joeyates/imap-backup/blob/feature/gmail-oauth. It's all documented [here](https://github.com/joeyates/imap-backup/blob/feature/gmail-oauth/docs/setting-up-gmail.md).

**Comment by paulgessinger** [comment]
Hey @joeyates, I actually don't have access to a paid Google account anymore, since I migrated hosts. (that was what my other issue had been about, actually :) ). 

So I'm afraid I can't test this I think.

**Comment by joeyates** [comment]
I've released [version 3.0.0](https://github.com/joeyates/imap-backup/releases/tag/v3.0.0) with OAuth2 support for GMail.
